connectors/source.py

```diff
@@ -105,20 +106,27 @@ def value(self):
     def value(self, value):
         self._value = value
 
-    def _convert(self, value, field_type):
-        if not isinstance(value, str):
-            # we won't convert the value if it's not a str
+    def _convert(self, value, field_type_):
+        cast_type = locate(field_type_)
+        if cast_type not in [str, int, float, bool, list]:
+            # unsupported type
             return value
 
-        if field_type == "int":
-            return int(value)
-        elif field_type == "float":
-            return float(value)
-        elif field_type == "bool":
-            return value.lower() in ("y", "yes", "true", "1")
-        elif field_type == "list":
-            return [item.strip() for item in value.split(",")]
-        return value
+        if isinstance(value, cast_type):  # pyright:ignore
+            return value
+
+        # list requires special type casting
+        if field_type_ == "list":
+            if isinstance(value, str):
+                return [item.strip() for item in value.split(",")] if value else []
+            else:
+                return [value] if value is not None else []
+
+        # strings don't get nullified
+        if value is None and field_type_ == "str":
+            return ""
+
+        return cast_type(value) if value else None  # pyright:ignore
```

Comment:
One suggestion, WDYT if we define a constant:
```python
TYPE_DEFAULTS = {
    str: "",
    int: None,
    float: None,
    bool: None,
    list: [],
}
```

So this can be simplified to
```suggestion
        if value is None:
            return TYPE_DEFAULTS[cast_type]

        return cast_type(value)  # pyright:ignore
```

At the beginning of the method, the unsupported type check can be simplified to:
```python
if cast_type not in TYPE_DEFAULTS:
            # unsupported type
            return value
```

One benefit is that if you want to add more type in the future, you only need to add to `TYPE_DEFAULTS`

Reply:
Nice idea, I've added this 🚀